src/lib/containers/UpsetPlot.tsx

```diff
@@ -0,0 +1,132 @@
+  maxBarCount?: number // will show all if not set
+  setName?: string // instead of "Set Size"
+  combinationName?: string // instead of "Intersection Size"
+  height?: number
+  loadingScreen: JSX.Element
+  token?: string
+}
+
+/**
+ * Upset plot.  See https://medium.com/@sgratzl/upset-js-javascript-tutorial-1b84bfd6896d
+ * Currently this is driven by a Synapse Table/View, but we may want to alter it to read in the data from a flat file instead.
+ */
+const UpsetPlot: React.FunctionComponent<UpsetPlotProps> = ({
+  entityId,
+  sql,
+  maxBarCount,
+  setName,
+  combinationName,
+  height = 700,
+  loadingScreen,
+  token,
+}) => {
+  const [isLoading, setIsLoading] = useState<boolean>()
+  const [sets, setSets] = useState<any>()
+  const [combinations, setCombinations] = useState<any>()
+  const [selection, setSelection] = React.useState(null as ISetLike<any> | null)
+
+  const { colorPalette } = getColorPallette(0, 1)
+  const updateFontSizes: UpSetFontSizes = {
+    setLabel: '14px'
+  }
+
+  useEffect(() => {
+    let isCancelled: boolean = false
+    const getPlotData = async () => {
+      setIsLoading(true)
+
+      const partMask = SynapseConstants.BUNDLE_MASK_QUERY_RESULTS
+      const queryRequest: QueryBundleRequest = {
+        partMask,
+        concreteType: 'org.sagebionetworks.repo.model.table.QueryBundleRequest',
+        entityId,
+        query: {
+          sql
+        },
+      }
+      const {
+        queryResult
+      } = await SynapseClient.getFullQueryTableResults(
+        queryRequest,
+        token
+      )
+      // transform query data into plot data, and store.
+      // collect all values for each key
+      const keyValuesMap = {}
+      // keyValuesMap looks like { 'A': {name: 'A', sets: ['S1', 'S2'] }, 'B': { name: 'B', sets: ['S1'] }, ... }
+      // It's a little redudant, but makes the next step much easier.
+      for (const row of queryResult.queryResults.rows) {
+        for (let j = 1; j < row.values.length; j += 1) {
+          const rowValues: any = row.values
+          const key = rowValues[0]
+          const newValue = rowValues[j]
+          keyValuesMap[key] = keyValuesMap[key] || {}
+          keyValuesMap[key].name = key
+
+          keyValuesMap[key].sets = keyValuesMap[key].sets || []
+          const found = keyValuesMap[key].sets.find((v: any) => v == newValue)
+          if (!found) {
+            keyValuesMap[key].sets.push(newValue)
+          }
+        }
+      }
+      // now create the expected elems set
+      const elems: any[] = Object.values(keyValuesMap)
+      // elems looks like [{ name: 'A', sets: ['S1', 'S2'] }, { name: 'B', sets: ['S1'] }, ...]
+      const sets = extractSets(elems);
+      const combinations = generateCombinations(sets, {
+        type: 'intersection',
+        min: 1,
+        limit: maxBarCount,
+        order: 'cardinality:desc',
+      })
+      if (!isCancelled) {
+        setSets(sets)
```

Comment:
For future reference, since the `sets` and `combinations` variables are set at the same time they could also be put into a single state variable, like so
`const [data, setData] = useState<{sets, combinations }: {sets: null as ISetLike<any> | null, combinations: any} >()`

The benefit of the approach above ^ would be that it would reduce the number of render calls (one versus two) and amount of child component updates. And just to be clear, the only reason I'm saying those two variable could be put into one is because they both get updated at the same time.

Reply:
updated to a single state variable (of new UpsetPlotData type)